// asyncc/lib/index.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "_setImmediate", {
  enumerable: true,
  get: function get() {
    return _setImmediate2._setImmediate;
  }
});
Object.defineProperty(exports, "AsynccError", {
  enumerable: true,
  get: function get() {
    return _AsynccError["default"];
  }
});
Object.defineProperty(exports, "compose", {
  enumerable: true,
  get: function get() {
    return _compose["default"];
  }
});
Object.defineProperty(exports, "connect", {
  enumerable: true,
  get: function get() {
    return _connect["default"];
  }
});
Object.defineProperty(exports, "doUntil", {
  enumerable: true,
  get: function get() {
    return _doUntil["default"];
  }
});
Object.defineProperty(exports, "doWhilst", {
  enumerable: true,
  get: function get() {
    return _doWhilst["default"];
  }
});
Object.defineProperty(exports, "each", {
  enumerable: true,
  get: function get() {
    return _each["default"];
  }
});
Object.defineProperty(exports, "eachLimit", {
  enumerable: true,
  get: function get() {
    return _eachLimit["default"];
  }
});
Object.defineProperty(exports, "eachSeries", {
  enumerable: true,
  get: function get() {
    return _eachSeries["default"];
  }
});
Object.defineProperty(exports, "NoPromise", {
  enumerable: true,
  get: function get() {
    return _NoPromise["default"];
  }
});
Object.defineProperty(exports, "noPromise", {
  enumerable: true,
  get: function get() {
    return _NoPromise.noPromise;
  }
});
Object.defineProperty(exports, "parallel", {
  enumerable: true,
  get: function get() {
    return _parallel["default"];
  }
});
Object.defineProperty(exports, "parallelLimit", {
  enumerable: true,
  get: function get() {
    return _parallelLimit["default"];
  }
});
Object.defineProperty(exports, "Queue", {
  enumerable: true,
  get: function get() {
    return _Queue["default"];
  }
});
Object.defineProperty(exports, "queue", {
  enumerable: true,
  get: function get() {
    return _Queue.queue;
  }
});
Object.defineProperty(exports, "retry", {
  enumerable: true,
  get: function get() {
    return _retry["default"];
  }
});
Object.defineProperty(exports, "series", {
  enumerable: true,
  get: function get() {
    return _series["default"];
  }
});
Object.defineProperty(exports, "times", {
  enumerable: true,
  get: function get() {
    return _times["default"];
  }
});
Object.defineProperty(exports, "until", {
  enumerable: true,
  get: function get() {
    return _until["default"];
  }
});
Object.defineProperty(exports, "whilst", {
  enumerable: true,
  get: function get() {
    return _whilst["default"];
  }
});
exports["default"] = void 0;

var _setImmediate2 = require("./_setImmediate");

var _AsynccError = _interopRequireDefault(require("./AsynccError"));

var _compose = _interopRequireDefault(require("./compose"));

var _connect = _interopRequireDefault(require("./connect"));

var _doUntil = _interopRequireDefault(require("./doUntil"));

var _doWhilst = _interopRequireDefault(require("./doWhilst"));

var _each = _interopRequireDefault(require("./each"));

var _eachLimit = _interopRequireDefault(require("./eachLimit"));

var _eachSeries = _interopRequireDefault(require("./eachSeries"));

var _NoPromise = _interopRequireWildcard(require("./NoPromise"));

var _parallel = _interopRequireDefault(require("./parallel"));

var _parallelLimit = _interopRequireDefault(require("./parallelLimit"));

var _Queue = _interopRequireWildcard(require("./Queue"));

var _retry = _interopRequireDefault(require("./retry"));

var _series = _interopRequireDefault(require("./series"));

var _times = _interopRequireDefault(require("./times"));

var _until = _interopRequireDefault(require("./until"));

var _whilst = _interopRequireDefault(require("./whilst"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

/**
* Serial execution patterns
* @module serial
*/

/**
* Parallel execution patterns
* @module parallel
*/
var _default = {
  _setImmediate: _setImmediate2._setImmediate,
  AsynccError: _AsynccError["default"],
  compose: _compose["default"],
  connect: _connect["default"],
  doUntil: _doUntil["default"],
  doWhilst: _doWhilst["default"],
  each: _each["default"],
  eachLimit: _eachLimit["default"],
  eachSeries: _eachSeries["default"],
  noPromise: _NoPromise.noPromise,
  NoPromise: _NoPromise["default"],
  parallel: _parallel["default"],
  parallelLimit: _parallelLimit["default"],
  queue: _Queue.queue,
  Queue: _Queue["default"],
  retry: _retry["default"],
  series: _series["default"],
  times: _times["default"],
  until: _until["default"],
  whilst: _whilst["default"]
};
exports["default"] = _default;
